File: rest-service/src/main/java/mb/api/service/http/report/ReportDownload.java
package mb.api.service.http.report;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.StreamingOutput;

import mb.lib.http.Header;
import mb.lib.http.MimeType;
import org.veupathdb.lib.blast.field.FormatType;

public class ReportDownload implements StreamingOutput
{
  private static final String AttachmentPat = "attachment; filename=\"%s\"";

  private final String      fileName;
  private final boolean     download;
  private final InputStream stream;

  public ReportDownload(
    String      fileName,
    boolean     download,
    InputStream stream
  ) {
    this.fileName = fileName;
    this.download = download;
    this.stream   = stream;
  }

  @Override
  public void write(OutputStream output) throws IOException, WebApplicationException {
    try (var b = new BufferedInputStream(stream)) {
      var n = 0;
      var a = new byte[8192];

      while ((n = b.read(a)) > 0)
        output.write(a, 0, n);
    }
  }

  public Response toResponse() {
    return configureResponse(Response.status(200)).entity(this).build();
  }

  private Response.ResponseBuilder configureResponse(Response.ResponseBuilder b) {
    String ct;

    if (fileName.endsWith(".xml")) {
      ct = MimeType.ApplicationXML;
    } else if (fileName.endsWith(".json")) {
      ct = MimeType.ApplicationJSON;
    } else if (fileName.endsWith(".zip")) {
      ct = MimeType.ApplicationZip;
    } else {
      ct = MimeType.TextPlain;
    }

    b = b.header(Header.ContentType, ct);

    if (download)
      b = b.header(Header.ContentDisposition, String.format(AttachmentPat, fileName));

    return b;
  }
}

File: rest-service/src/main/kotlin/mb/api/service/http/report/ReportDownload.kt
package mb.api.service.http.report

import jakarta.ws.rs.core.Response
import jakarta.ws.rs.core.StreamingOutput
import mb.api.service.util.RangedStream
import mb.lib.http.Header
import mb.lib.http.MimeType
import java.io.OutputStream

internal data class ReportDownload(
  private val fileName: String,
  private val download: Boolean,
  private val stream:   RangedStream,
): StreamingOutput {

  companion object {
    private const val AttachmentPat = "attachment; filename=\"%s\""
  }

  override fun write(output: OutputStream) {
    stream.write(output)
  }

  fun toResponse(): Response {
    return configureResponse(Response.status(200)).entity(this).build()
  }

  private fun configureResponse(b: Response.ResponseBuilder): Response.ResponseBuilder {
    b.header(Header.ContentType, when {
      fileName.endsWith(".xml")  -> MimeType.ApplicationXML
      fileName.endsWith(".json") -> MimeType.ApplicationJSON
      fileName.endsWith(".zip")  -> MimeType.ApplicationZip
      else                       -> MimeType.TextPlain
    })

    if (download)
      b.header(Header.ContentDisposition, String.format(AttachmentPat, fileName))

    return b
  }
}
